src/client/serialize/daos_serialize.c

```diff
@@ -1175,3 +1252,1807 @@ daos_cont_deserialize_attrs(char *filename, uint64_t *num_attrs,
+	daos_handle_t		oh;
+	uint32_t		oids_nr;
+	bool			is_kv = false;
+	float			version = 0.0;
+	hid_t			usr_attr_memtype = -1;
+	hid_t			usr_attr_name_vtype = -1;
+	hid_t			usr_attr_val_vtype = -1;
+	struct oid		*oid_tmp = NULL;
+
+	if (filename == NULL)
+		D_GOTO(out, rc = -DER_INVAL);
+
+	/* Create HDF5 file for serialization */
+	hdf5.file = H5Fcreate(filename, H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
+	if (hdf5.file < 0) {
+		D_ERROR("Failed to create HDF5 file: %s\n", filename);
+		D_GOTO(out, rc = -DER_IO);
+	}
+
+	D_PRINT("Serializing Container to: %s\n", filename);
+
+	/* serialize cont version */
+	rc = serialize_version(&hdf5, version);
+	if (rc != 0) {
+		D_ERROR("Failed to serialize version\n");
+		D_GOTO(out, rc);
+	}
+
+	/* serialize cont props */
+	rc = daos_cont_serialize_props(hdf5.file, props);
+	if (rc != 0) {
+		D_ERROR("failed to serialize cont layout "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out, rc);
+	}
+
+	/* serialize usr_attrs if there are any */
+	if (num_attrs > 0) {
+		/* create User Attributes Dataset for file */
+		usr_attr_memtype = H5Tcreate(H5T_COMPOUND,
+					     sizeof(struct usr_attr));
+		if (usr_attr_memtype < 0) {
+			D_ERROR("failed to create memory datatype\n");
+			D_GOTO(out, rc = -DER_MISC);
+		}
+
+		usr_attr_name_vtype = H5Tcopy(H5T_C_S1);
+		if (usr_attr_name_vtype < 0) {
+			D_ERROR("failed to create variable datatype\n");
+			D_GOTO(out, rc = -DER_MISC);
+		}
+		status = H5Tset_size(usr_attr_name_vtype, H5T_VARIABLE);
+		if (status < 0) {
+			D_ERROR("failed to set datatype size\n");
+			D_GOTO(out, rc = -DER_MISC);
+		}
+		usr_attr_val_vtype = H5Tvlen_create(H5T_NATIVE_OPAQUE);
+		if (usr_attr_val_vtype < 0) {
+			D_ERROR("failed to variable length type\n");
+			D_GOTO(out, rc = -DER_MISC);
+		}
+		status = H5Tinsert(usr_attr_memtype, "Attribute Name",
+				   HOFFSET(struct usr_attr, attr_name),
+				   usr_attr_name_vtype);
+		if (status < 0) {
+			D_ERROR("failed to insert into compound datatype\n");
+			D_GOTO(out, rc = -DER_MISC);
+		}
+		status = H5Tinsert(usr_attr_memtype, "Attribute Value",
+				   HOFFSET(struct usr_attr, attr_val),
+				   usr_attr_val_vtype);
+		if (status < 0) {
+			D_ERROR("failed to insert into compound datatype\n");
+			D_GOTO(out, rc = -DER_MISC);
+		}
+		rc = daos_cont_serialize_attrs(hdf5.file, &usr_attr_memtype,
+					       num_attrs, names, buffers,
+					       sizes);
+		if (rc != 0) {
+			D_ERROR("failed to serialize usr attributes "DF_RC"\n",
+				DP_RC(rc));
+			D_GOTO(out, rc);
+		}
+	}
+
+	/* create oid, dkey, akey buffers for container data */
+	D_ALLOC(hdf5.oid_data, sizeof(struct oid));
+	if (hdf5.oid_data == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+	D_ALLOC(hdf5.dkey_data, sizeof(struct dkey));
+	if (hdf5.dkey_data == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+	D_ALLOC(hdf5.akey_data, sizeof(struct akey));
+	if (hdf5.akey_data == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+	hdf5.oid = &(hdf5.oid_data);
+	hdf5.dk = &(hdf5.dkey_data);
+	hdf5.ak = &(hdf5.akey_data);
```

Comment:
im not understanding why all this is required since all those are member of the struct

Reply:
Agree. They seemed to just be a shorthand to the first element in each array. But then that shorthand was being dereferenced anyway... So I re-stitched to use the *_data members directly